In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import pandas as pd
path= '/content/drive/MyDrive/Colab aprendizaje supervisado/Taller 2/Real estate.csv'
df= pd.read_csv(path)

In [4]:
df.head()

,No,X1 transaction date,X2 house age,X3 distance to the nearest MRT station,X4 number of convenience stores,X5 latitude,X6 longitude,Y house price of unit area
0,1,2012.917,32.0,84.87882,10,24.98298,121.54024,37.9
1,2,2012.917,19.5,306.59470,9,24.98034,121.53951,42.2
2,3,2013.583,13.3,561.98450,5,24.98746,121.54391,47.3
3,4,2013.500,13.3,561.98450,5,24.98746,121.54391,54.8
4,5,2012.833,5.0,390.56840,5,24.97937,121.54245,43.1


In [5]:
df.describe()

,No,X1 transaction date,X2 house age,X3 distance to the nearest MRT station,X4 number of convenience stores,X5 latitude,X6 longitude,Y house price of unit area
count,414.000000,414.000000,414.000000,414.000000,414.000000,414.000000,414.000000,414.000000
mean,207.500000,2013.148971,17.712560,1083.885689,4.094203,24.969030,121.533361,37.980193
std,119.655756,0.281967,11.392485,1262.109595,2.945562,0.012410,0.015347,13.606488
min,1.000000,2012.667000,0.000000,23.382840,0.000000,24.932070,121.473530,7.600000
25%,104.250000,2012.917000,9.025000,289.324800,1.000000,24.963000,121.528085,27.700000
50%,207.500000,2013.167000,16.100000,492.231300,4.000000,24.971100,121.538630,38.450000
75%,310.750000,2013.417000,28.150000,1454.279000,6.000000,24.977455,121.543305,46.600000
max,414.000000,2013.583000,43.800000,6488.021000,10.000000,25.014590,121.566270,117.500000


In [6]:
df.shape

(414, 8)

In [7]:
df.isnull().sum()

,0
No,0
X1 transaction date,0
X2 house age,0
X3 distance to the nearest MRT station,0
X4 number of convenience stores,0
X5 latitude,0
X6 longitude,0
Y house price of unit area,0


In [8]:
df = df.drop('No', axis=1)

## Partición entrenamiento y prueba

In [9]:
train= df.sample(frac=0.7, random_state=42)
test= df.drop(train.index)

In [10]:
train.head()

,X1 transaction date,X2 house age,X3 distance to the nearest MRT station,X4 number of convenience stores,X5 latitude,X6 longitude,Y house price of unit area
358,2013.167,1.1,193.5845,6,24.96571,121.54089,45.1
350,2013.000,13.2,492.2313,5,24.96515,121.53737,42.3
373,2013.083,0.0,274.0144,1,24.97480,121.53059,52.2
399,2012.917,12.7,170.1289,1,24.97371,121.52984,37.3
369,2012.667,20.2,2185.1280,3,24.96322,121.51237,22.8


In [11]:
from sklearn.model_selection import train_test_split

X = df.drop('Y house price of unit area', axis=1)
y = df['Y house price of unit area']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [12]:
df_train = pd.concat([y_train, X_train], axis=1)
df_train.head()

,Y house price of unit area,X1 transaction date,X2 house age,X3 distance to the nearest MRT station,X4 number of convenience stores,X5 latitude,X6 longitude
7,46.7,2013.417,20.3,287.60250,6,24.98042,121.54228
194,29.3,2013.500,15.2,3771.89500,0,24.93363,121.51158
19,47.7,2012.667,1.5,23.38284,7,24.96772,121.54102
176,19.2,2012.833,13.9,4573.77900,0,24.94867,121.49507
367,20.9,2012.833,15.0,1828.31900,2,24.96464,121.51531


In [13]:
corr_matrix = df_train.corr()
corr_matrix

,Y house price of unit area,X1 transaction date,X2 house age,X3 distance to the nearest MRT station,X4 number of convenience stores,X5 latitude,X6 longitude
Y house price of unit area,1.000000,0.081105,-0.204978,-0.678174,0.574187,0.549832,0.526084
X1 transaction date,0.081105,1.000000,0.013485,0.080023,-0.028160,0.022414,-0.056164
X2 house age,-0.204978,0.013485,1.000000,0.046619,0.020733,0.026494,-0.069503
X3 distance to the nearest MRT station,-0.678174,0.080023,0.046619,1.000000,-0.610834,-0.559727,-0.820703
X4 number of convenience stores,0.574187,-0.028160,0.020733,-0.610834,1.000000,0.434704,0.465457
X5 latitude,0.549832,0.022414,0.026494,-0.559727,0.434704,1.000000,0.408543
X6 longitude,0.526084,-0.056164,-0.069503,-0.820703,0.465457,0.408543,1.000000


In [14]:
X_train = X_train.drop('X6 longitude', axis=1)
X_test = X_test.drop('X6 longitude', axis=1)

In [15]:
import statsmodels.api as sm

# Intercepto
X_const = sm.add_constant(X_train)

# Modelo de regresión
modelo_sm = sm.OLS(y_train, X_const).fit()

# Imprimir el resumen del modelo (incluye la ecuación y ANOVA)
print(modelo_sm.summary())

# Obtener y mostrar los residuales
residuales = modelo_sm.resid
print("\nResiduales:")
print(residuales)

                                OLS Regression Results                                
Dep. Variable:     Y house price of unit area   R-squared:                       0.585
Model:                                    OLS   Adj. R-squared:                  0.578
Method:                         Least Squares   F-statistic:                     79.84
Date:                        Sat, 06 Sep 2025   Prob (F-statistic):           4.91e-52
Time:                                21:27:28   Log-Likelihood:                -1042.3
No. Observations:                         289   AIC:                             2097.
Df Residuals:                             283   BIC:                             2119.
Df Model:                                   5                                         
Covariance Type:                    nonrobust                                         
                                             coef    std err          t      P>|t|      [0.025      0.975]
-----------------------

In [16]:
def backward_elimination(X_train, y_train):

    selected_features = list(X_train.columns)
    X_temp_initial = sm.add_constant(X_train[selected_features])
    model_initial = sm.OLS(y_train, X_temp_initial).fit()
    best_aic = model_initial.aic
    print(f"AIC inicial with all features: {best_aic:.2f}")

    while selected_features:
        feature_to_remove = None
        improved = False
        worst_aic = best_aic # Inicializar con el mejor AIC

        for feature in selected_features:
            temp_features = [f for f in selected_features if f != feature]
            if not temp_features:
                 X_temp = pd.DataFrame({'const': np.ones(len(X_train))})
            else:
                X_temp = sm.add_constant(X_train[temp_features])

            model = sm.OLS(y_train, X_temp).fit()
            current_aic = model.aic

            # En backward elimination, se eliminan las variables que no están, cuando se obtuvo el mejor AIC
            if current_aic < best_aic:
                best_aic = current_aic
                feature_to_remove = feature
                improved = True

        if improved:
            selected_features.remove(feature_to_remove)
            print(f"Variables removidas {feature_to_remove}. Nuevo mejor AIC: {best_aic:.2f}")
        else:
            break

    return selected_features

selected_features_backward = backward_elimination(X_train, y_train)
print("\nVariables seleccionadas usando backward elimination:", selected_features_backward)

AIC inicial with all features: 2096.58

Variables seleccionadas usando backward elimination: ['X1 transaction date', 'X2 house age', 'X3 distance to the nearest MRT station', 'X4 number of convenience stores', 'X5 latitude']


In [17]:
from sklearn.metrics import mean_squared_error, mean_absolute_error
import numpy as np

# Calculate predictions
X_test_const = sm.add_constant(X_test[selected_features_backward])
predictions = modelo_sm.predict(X_test_const)

# Calculate metrics
r2 = modelo_sm.rsquared
adj_r2 = modelo_sm.rsquared_adj
rmse = np.sqrt(mean_squared_error(y_test, predictions))
mae = mean_absolute_error(y_test, predictions)
# Avoid division by zero in MAPE
mape = np.mean(np.abs((y_test - predictions) / y_test)) * 100

print(f"R2: {r2:.2f}")
print(f"Adjusted R2: {adj_r2:.2f}")
print(f"RMSE: {rmse:.2f}")
print(f"MAE: {mae:.2f}")
print(f"MAPE: {mape:.2f}%")

R2: 0.59
Adjusted R2: 0.58
RMSE: 8.53
MAE: 6.14
MAPE: 21.66%


In [21]:
current_aic = modelo_sm.aic
current_bic = modelo_sm.bic
print("aic:", current_aic)
print("bic:", current_bic)

aic: 2096.57907713374
bic: 2118.5776372624146


Eliminé la variable X6 longitude, porque tiene una correlación importante a la variable X5 Latitude, además porque al comparar ambas con Y, Latitude tiene una correlación mayor.

Dejé la variable Transaction Date, porque aunque su correlación lineal es baja, el precio de las viviendas suele estar influenciado por el momento de la transacción (ciclos económicos, inflación, burbujas inmobiliarias).

## Regresión regularizada

In [23]:
from sklearn.linear_model import Ridge, Lasso, ElasticNet
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import numpy as np

# Define the features (X) and target (y)
# Using all features from X_train and X_test as SFS is not defined
X_train_regularized = X_train
X_test_regularized = X_test

# Add a small epsilon to prevent division by zero in MAPE
epsilon = 1e-10

# Ridge Regression
ridge_model = Ridge(alpha=1.0) # You can tune alpha
ridge_model.fit(X_train_regularized, y_train)
y_pred_ridge = ridge_model.predict(X_test_regularized)

print("--- Ridge Regression ---")
print(f"RMSE: {np.sqrt(mean_squared_error(y_test, y_pred_ridge)):.2f}")
print(f"MAE: {mean_absolute_error(y_test, y_pred_ridge):.2f}")
mape_ridge = np.mean(np.abs((y_test - y_pred_ridge) / (y_test + epsilon))) * 100
print(f"MAPE: {mape_ridge:.2f}%")
print(f"R2: {r2_score(y_test, y_pred_ridge):.2f}")

# Lasso Regression
lasso_model = Lasso(alpha=1.0) # You can tune alpha
lasso_model.fit(X_train_regularized, y_train)
y_pred_lasso = lasso_model.predict(X_test_regularized)

print("\n--- Lasso Regression ---")
print(f"RMSE: {np.sqrt(mean_squared_error(y_test, y_pred_lasso)):.2f}")
print(f"MAE: {mean_absolute_error(y_test, y_pred_lasso):.2f}")
mape_lasso = np.mean(np.abs((y_test - y_pred_lasso) / (y_test + epsilon))) * 100
print(f"MAPE: {mape_lasso:.2f}%")
print(f"R2: {r2_score(y_test, y_pred_lasso):.2f}")

# Elastic Net Regression
elastic_net_model = ElasticNet(alpha=1.0, l1_ratio=0.5) # You can tune alpha and l1_ratio
elastic_net_model.fit(X_train_regularized, y_train)
y_pred_elastic_net = elastic_net_model.predict(X_test_regularized)

print("\n--- Elastic Net Regression ---")
print(f"RMSE: {np.sqrt(mean_squared_error(y_test, y_pred_elastic_net)):.2f}")
print(f"MAE: {mean_absolute_error(y_test, y_pred_elastic_net):.2f}")
mape_elastic_net = np.mean(np.abs((y_test - y_pred_elastic_net) / (y_test + epsilon))) * 100
print(f"MAPE: {mape_elastic_net:.2f}%")
print(f"R2: {r2_score(y_test, y_pred_elastic_net):.2f}")

print("--- Linear Regression Model Metrics (Test Set) ---")
print(f"RMSE: {rmse:.2f}")
print(f"MAE: {mae:.2f}")
print(f"R2: {r2:.2f}")
print(f"Adjusted R2: {adj_r2:.2f}")
print(f"MAPE: {mape:.2f}%")

--- Ridge Regression ---
RMSE: 8.63
MAE: 6.33
MAPE: 22.62%
R2: 0.56

--- Lasso Regression ---
RMSE: 8.73
MAE: 6.37
MAPE: 22.99%
R2: 0.54

--- Elastic Net Regression ---
RMSE: 8.73
MAE: 6.36
MAPE: 22.98%
R2: 0.54
--- Linear Regression Model Metrics (Test Set) ---
RMSE: 8.53
MAE: 6.14
R2: 0.59
Adjusted R2: 0.58
MAPE: 21.66%


In [22]:
import numpy as np
from sklearn.metrics import mean_squared_error

# Function to calculate approximate AIC and BIC for scikit-learn models
def calculate_aic_bic(n, mse, k):
    """
    Calculates approximate AIC and BIC.

    Args:
        n (int): Number of observations.
        mse (float): Mean squared error of the model.
        k (int): Number of parameters (including intercept).

    Returns:
        tuple: (AIC, BIC)
    """
    # Approximate log-likelihood based on MSE
    log_likelihood = -n / 2 * np.log(2 * np.pi * mse) - n / 2

    # AIC = -2 * log_likelihood + 2 * k
    aic = -2 * log_likelihood + 2 * k

    # BIC = -2 * log_likelihood + k * log(n)
    bic = -2 * log_likelihood + k * np.log(n)

    return aic, bic

n_train = X_train_regularized.shape[0]
k_ridge = X_train_regularized.shape[1] + 1 # +1 for intercept
k_lasso = X_train_regularized.shape[1] + 1 # +1 for intercept
k_elastic_net = X_train_regularized.shape[1] + 1 # +1 for intercept


# Calculate MSE on the training data for each model
y_pred_train_ridge = ridge_model.predict(X_train_regularized)
mse_train_ridge = mean_squared_error(y_train, y_pred_train_ridge)

y_pred_train_lasso = lasso_model.predict(X_train_regularized)
mse_train_lasso = mean_squared_error(y_train, y_pred_train_lasso)

y_pred_train_elastic_net = elastic_net_model.predict(X_train_regularized)
mse_train_elastic_net = mean_squared_error(y_train, y_pred_train_elastic_net)


# Calculate AIC and BIC for each model
aic_ridge, bic_ridge = calculate_aic_bic(n_train, mse_train_ridge, k_ridge)
aic_lasso, bic_lasso = calculate_aic_bic(n_train, mse_train_lasso, k_lasso)
aic_elastic_net, bic_elastic_net = calculate_aic_bic(n_train, mse_train_elastic_net, k_elastic_net)

print("--- Approximate AIC and BIC ---")
print(f"Linear Regression: AIC={current_aic:.2f}, BIC={current_bic:.2f}")
print(f"Ridge: AIC={aic_ridge:.2f}, BIC={bic_ridge:.2f}")
print(f"Lasso: AIC={aic_lasso:.2f}, BIC={bic_lasso:.2f}")
print(f"Elastic Net: AIC={aic_elastic_net:.2f}, BIC={bic_elastic_net:.2f}")

--- Approximate AIC and BIC ---
Linear Regression: AIC=2096.58, BIC=2118.58
Ridge: AIC=2117.01, BIC=2139.01
Lasso: AIC=2129.88, BIC=2151.88
Elastic Net: AIC=2129.80, BIC=2151.80


## Coeficientes de cada uno de los modelos de regresión con regularización

In [24]:
print("\n--- Linear Regression Coefficients ---")
print("Intercept (const):", modelo_sm.params['const'])
for feature, coef in zip(X_train_regularized.columns, modelo_sm.params[1:]):
    print(f"{feature}: {coef:.4f}")

print("\n--- Ridge Regression Coefficients ---")
for feature, coef in zip(X_train_regularized.columns, ridge_model.coef_):
    print(f"{feature}: {coef:.4f}")

print("\n--- Lasso Regression Coefficients ---")
for feature, coef in zip(X_train_regularized.columns, lasso_model.coef_):
    print(f"{feature}: {coef:.4f}")

print("\n--- Elastic Net Regression Coefficients ---")
for feature, coef in zip(X_train_regularized.columns, elastic_net_model.coef_):
    print(f"{feature}: {coef:.4f}")


--- Linear Regression Coefficients ---
Intercept (const): -17719.61096160066
X1 transaction date: 5.8050
X2 house age: -0.2408
X3 distance to the nearest MRT station: -0.0046
X4 number of convenience stores: 1.0866
X5 latitude: 243.3418

--- Ridge Regression Coefficients ---
X1 transaction date: 6.2056
X2 house age: -0.2290
X3 distance to the nearest MRT station: -0.0057
X4 number of convenience stores: 1.2276
X5 latitude: 7.1335

--- Lasso Regression Coefficients ---
X1 transaction date: 0.0000
X2 house age: -0.2173
X3 distance to the nearest MRT station: -0.0059
X4 number of convenience stores: 1.0696
X5 latitude: 0.0000

--- Elastic Net Regression Coefficients ---
X1 transaction date: 0.0300
X2 house age: -0.2203
X3 distance to the nearest MRT station: -0.0059
X4 number of convenience stores: 1.0645
X5 latitude: 0.0000


## Regresión lineal con variables transaction date y number of convenience stores

In [25]:
import statsmodels.api as sm

# Select the desired features
X_train_selected = X_train[['X1 transaction date', 'X4 number of convenience stores']]
y_train_selected = y_train

# Add a constant for the intercept
X_train_selected_const = sm.add_constant(X_train_selected)

# Fit the OLS model
model_selected = sm.OLS(y_train_selected, X_train_selected_const).fit()

# Print the model summary
print(model_selected.summary())

                                OLS Regression Results                                
Dep. Variable:     Y house price of unit area   R-squared:                       0.339
Model:                                    OLS   Adj. R-squared:                  0.335
Method:                         Least Squares   F-statistic:                     73.39
Date:                        Sat, 06 Sep 2025   Prob (F-statistic):           1.88e-26
Time:                                21:40:17   Log-Likelihood:                -1109.6
No. Observations:                         289   AIC:                             2225.
Df Residuals:                             286   BIC:                             2236.
Df Model:                                   2                                         
Covariance Type:                    nonrobust                                         
                                      coef    std err          t      P>|t|      [0.025      0.975]
------------------------------

De acuerdo a los coeficientes de los modelos de regresión con regularización, es posible realizar el modelo usando sólo dos variables (Transaction Date y Number of Convenience Stores), sin embargo, las métricas de desempeño del nuevo modelo indican que reducir las variables independientes a estas dos no mejoran los resultados del modelo anterior, en su lugar son peores.

Por lo tanto, escogeré el modelo anterior.

# 📊 Informe Ejecutivo y Técnico – Predicción de Precios de Vivienda

## 1. Objetivo del análisis
El propósito del estudio fue comparar distintos enfoques de regresión estadística y de machine learning para **predecir el precio de vivienda por unidad de área**. Además, se buscó identificar las variables más determinantes en el valor de los inmuebles y traducir los hallazgos en **conclusiones de negocio** que apoyen decisiones estratégicas en el sector inmobiliario.

---

## 2. Metodología

1. **Dataset**:  
   - 414 registros de transacciones inmobiliarias.  
   - Variables:  
     - `X1 transaction date`: fecha de transacción (indicador temporal).  
     - `X2 house age`: antigüedad de la vivienda.  
     - `X3 distance to the nearest MRT station`: distancia a la estación de metro más cercana.  
     - `X4 number of convenience stores`: número de tiendas cercanas.  
     - `X5 latitude` y `X6 longitude`: coordenadas geográficas.  
     - `Y house price of unit area`: precio por metro cuadrado (variable objetivo).

2. **Modelos comparados**:  
   - Regresión Lineal (baseline).  
   - Ridge Regression (regularización L2).  
   - Lasso Regression (regularización L1, selección de variables).  
   - Elastic Net (combinación de L1 y L2).

3. **Criterios de evaluación**:  
   - Métricas de error: RMSE, MAE, MAPE.  
   - Bondad de ajuste: R² y R² ajustado.  
   - Criterios de información: AIC y BIC (parcimonia del modelo).

---

## 3. Resultados cuantitativos

| Modelo | RMSE | MAE | MAPE | R² | Adj R² | AIC | BIC |
|--------|------|-----|------|----|--------|-----|-----|
| **Regresión Lineal** | **8.53** | **6.14** | **21.66%** | **0.59** | 0.58 | **2096.58** | **2118.58** |
| Ridge | 8.63 | 6.33 | 22.62% | 0.56 | – | 2117.01 | 2139.01 |
| Lasso | 8.73 | 6.37 | 22.99% | 0.54 | – | 2129.88 | 2151.88 |
| Elastic Net | 8.73 | 6.36 | 22.98% | 0.54 | – | 2129.80 | 2151.80 |

📌 **Hallazgo técnico clave**:  
El modelo lineal simple obtiene **mejores métricas de error, mayor R² y mejores valores de AIC/BIC**, lo que sugiere que el dataset no presenta alta multicolinealidad ni necesidad de regularización fuerte.

---

## 4. Análisis de coeficientes

### 4.1 Regresión Lineal (OLS)
- `X4 number of convenience stores`: coeficiente positivo y altamente significativo (**+2.67, p<0.01**). Cada tienda adicional cercana aumenta en promedio el valor del m².  
- `X1 transaction date`: coeficiente positivo (**+4.77, p=0.044**), indicando **tendencia al alza de precios en el tiempo**.  
- `X2 house age`: coeficiente negativo (-0.24), lo que refleja la pérdida de valor con la antigüedad.  
- `X3 distance to MRT`: coeficiente pequeño y no significativo, sugiriendo poco impacto en este contexto.  
- `X5 latitude` y `X6 longitude`: significativos en el modelo lineal, pero eliminados por Lasso → su efecto es débil y posiblemente redundante.

### 4.2 Regularización
- **Lasso** anuló las variables `transaction date` y `latitude`, confirmando que su efecto no es robusto en presencia de otras variables.  
- **Ridge** y **Elastic Net** redujeron coeficientes pero no mejoraron la precisión.

---

## 5. Implicaciones de negocio

1. **Cercanía a servicios comerciales**  
   - Es el factor más relevante para la valorización de las viviendas.  
   - Estrategia: priorizar desarrollos en zonas con alta densidad de tiendas y servicios básicos.  

2. **Tendencia temporal positiva**  
   - El mercado inmobiliario presenta una clara **apreciación a lo largo del tiempo**.  
   - Estrategia: proyectos adquiridos y vendidos en años recientes capturan mejor valorización.  

3. **Edad de la vivienda como depreciador**  
   - La antigüedad reduce el precio, aunque de manera menos marcada que los servicios.  
   - Estrategia: programas de remodelación o reposición pueden contrarrestar esta pérdida de valor.  

4. **Transporte público**  
   - La distancia al metro no resulta determinante en este dataset.  
   - Implica que en este mercado específico los compradores valoran más la conveniencia cotidiana (tiendas, servicios) que la conectividad.  

5. **Ubicación geográfica (lat/long)**  
   - Importan en el modelo simple, pero no son robustas frente a regularización.  
   - Esto indica que **la ubicación por sí sola no explica tanto como la cercanía a servicios específicos**.

---

## 6. Conclusiones

- **El modelo lineal simple es el más adecuado**, explicando alrededor del **59% de la variación en precios**.  
- **El principal motor de valorización es la densidad de tiendas cercanas**, seguido por el factor tiempo (mercado al alza).  
- **La edad de la vivienda impacta negativamente**, pero en menor medida.  
- **El transporte público no es un factor decisivo en este caso**, lo que orienta a estrategias basadas más en servicios comerciales que en movilidad.  
- Para futuras predicciones, incluir variables macroeconómicas (PIB, tasas de interés, inflación) podría mejorar la capacidad predictiva.

---
